include/libopencm3/stm32/h7/rcc.h

```diff
@@ -181,23 +328,228 @@ LGPL License Terms @ref lgpl_license
+#define RCC_D2CCIP1R_SPDIFSEL_PLL2R     0x1
+#define RCC_D2CCIP1R_SPDIFSEL_PLL3R     0x2
+#define RCC_D2CCIP1R_SPDIFSEL_HSI       0x3
+#define RCC_D2CCIP1R_SPI45SEL_APB4      0x0
+#define RCC_D2CCIP1R_SPI45SEL_PLL2Q     0x1
+#define RCC_D2CCIP1R_SPI45SEL_PLL3Q     0x2
+#define RCC_D2CCIP1R_SPI45SEL_HSI       0x3
+#define RCC_D2CCIP1R_SPI45SEL_CSI       0x4
+#define RCC_D2CCIP1R_SPI45SEL_HSE       0x5
+#define RCC_D2CCIP1R_SPI45SEL_MASK      0x7
+#define RCC_D2CCIP1R_SPI123SEL_PLL1Q    0x0
+#define RCC_D2CCIP1R_SPI123SEL_PLL2P    0x1
+#define RCC_D2CCIP1R_SPI123SEL_PLL3P    0x2
+#define RCC_D2CCIP1R_SPI123SEL_I2SCKIN  0x3
+#define RCC_D2CCIP1R_SPI123SEL_PERCK    0x4
+#define RCC_D2CCIP1R_SPI123SEL_MASK     0x7
+#define RCC_D2CCIP1R_SAISEL_PLL1Q       0x0
+#define RCC_D2CCIP1R_SAISEL_PLL2P       0x1
+#define RCC_D2CCIP1R_SAISEL_PLL3P       0x2
+#define RCC_D2CCIP1R_SAISEL_I2SCKIN     0x3
+#define RCC_D2CCIP1R_SAISEL_PERCK       0x4
+#define RCC_D2CCIP1R_SAISEL_MASK        0x7
+/**@}*/
+
+/** @defgroup rcc_d2ccip2r_values RCC_D2CCIP2R Values
+ * @ingroup rcc_registers
+ * @{*/
+#define RCC_D2CCIP2R_LPTIM1SEL_SHIFT        28
+#define RCC_D2CCIP2R_CECSEL_SHIFT           22
+#define RCC_D2CCIP2R_USBSEL_SHIFT           20
+#define RCC_D2CCIP2R_I2C123SEL_SHIFT        12
+#define RCC_D2CCIP2R_RNGSEL_SHIFT           8
+#define RCC_D2CCIP2R_USART16SEL_SHIFT       3
+#define RCC_D2CCIP2R_USART234578SEL_SHIFT   0
+
+#define RCC_D2CCIP2R_USART16SEL_PCLK2       0
+#define RCC_D2CCIP2R_USART234578SEL_PCLK1   0
+#define RCC_D2CCIP2R_USARTSEL_PLL2Q         1
+#define RCC_D2CCIP2R_USARTSEL_PLL3Q         2
+#define RCC_D2CCIP2R_USARTSEL_HSI           3
+#define RCC_D2CCIP2R_USARTSEL_CSI           4
+#define RCC_D2CCIP2R_USARTSEL_LSE           5
+#define RCC_D2CCIP2R_USARTSEL_MASK          7
+/**@}*/
+
+
+#define HSI_BASE_FREQUENCY              64000000UL
+
+/** Enumerations for clocks in the clock tree to allow user to get the current configuration of the
+ *  clocks from the RCC module. These clock sources will each be tracked through the settings.
+ */
+enum rcc_clock_source {
+  RCC_CPUCLK,
+  RCC_SYSCLK,
+  RCC_PERCLK,
+  RCC_SYSTICKCLK,
+  RCC_HCLK3,
+  RCC_AHBCLK,           /* AHB1,2,4 all share base HCLK. */
+  RCC_APB1CLK,          /* APB1 and PCLK1 are aliases. */
+  RCC_PCLK1,
+  RCC_APB2CLK,
+  RCC_PCLK2,
+  RCC_APB3CLK,
+  RCC_PCLK3,
+  RCC_APB4CLK,
+  RCC_PCLK4,
+  RCC_FDCAN1CLK,        /* FDCAN controllers share a kernel clock */
+  RCC_FDCAN2CLK,
+  RCC_SPI1CLK,          /* SPI 1/2/3 share a kernel clock. */
+  RCC_SPI2CLK,
+  RCC_SPI3CLK,
+  RCC_SPI4CLK,          /* SPI 4/5 share kernel clock. */
+  RCC_SPI5CLK,
+  RCC_SPI6CLK,          /* SPI6 has its own kernel clock. */
+  RCC_USART1CLK,        /* USART 1/6 share kernel clock. */
+  RCC_USART6CLK,
+  RCC_USART2CLK,        /* USART 2/3/4/5/6/8 share kernel clock. */
+  RCC_USART3CLK,
+  RCC_USART4CLK,
+  RCC_USART5CLK,
+  RCC_USART7CLK,
+  RCC_USART8CLK,
+};
```

Comment:
I'm a little concerned about how close this is to the bit macros used for enabling peripheral clocks, and being rather confusing.  We've also never (so far) really provided ways of looking up current clocks _back again_, we've normally considered that the user set the clock, providing backwrds functions wasn't really useful/necessary.  What do you see as the use case for these?

Reply:
I'm happy to rename these, but we have already proven the use case that this is pretty valuable in our H7 development. In short, this is the first time there has been a clock tree this complicated, and it makes pushing some of the work in clock discovery into the rcc implementation desirable. 

To do this without an actual tree discovery, the configuration structure has to hold all possible configurations of every clock selection in the system, and/or you need a global version of every clock in the system (which anyone can modify BTW). This makes the configuration kinda gross and unwieldy to get something working, plus the user then needs to know how to calculate each of these clocks by delving through the clock tree themselves. 

By breaking this up into a basic core set of configuration, just for the inputs and main PLLs, and then providing entry points for the clock tree variations, this makes setup a lot easier to manage (building blocks...core stuff works, then add my other clock selections). 

Clock discovery is not absolutely required, but as mentioned above, it provides encapsulation of the values, and it allows some optimization in storage since each value is not individually stored.

Reply:
I'd really like to see some example code of this, not just the headers if possible.  We have discussed in the past that rather than just keeping three clock uint32s, (ahb/apb1/apb2) to keep the clock struct itself, and I do know that the "choose a clock for this peripheral!" makes things baloon out,

Reply:
Here is a sample of the PLL configuration for our platform...
```c++
// ================================================================================================
static void InitLdoAndPll() {
  // Setup LDO for high speed operation.
  rcc_periph_clock_enable(RCC_SYSCFG);
  pwr_set_mode_ldo();                  // Ensure we're running on the LDO.
  pwr_set_vos_scale(PWR_VOS_SCALE_0);  // Set highest voltage output for 400+MHz.

  // Clock configuration for this platform.
  // clang-format off
  static const struct rcc_pll_config pll_config = {
    .hse_frequency = 8000000UL,
    .sysclk_mux    = RCC_SYSCLK_PLL,
    .pll_mux       = RCC_PLL_HSE,
    .pll1 = {
        .divm = 4U,     // 2MHz PLL1 base clock pre-multiplier, cancels post div2.
        .divn = 400U,   // 800MHz for PLL1 to drive the sysclk throu post-div of 2.
        .divp = 2U,     // PLL1P post-divider gives 400MHz output.
        .divq = 10U,    // PLL1Q post-divider gives 80MHz output for FDCAN.
                        // PLL1R are disabled for now.
    },
    .pll2 = {},
    .pll3 = {},
    // Set CPU to PLL1P output at 400MHz.
    .d1cfg_core_prescale  = RCC_D1CFGR_D1CPRE(RCC_D1CFGR_D1CPRE_BYP),
    // Constrain HPRE below 240MHz by dividing core clock by 2.
    .d1cfg_hclk3_prescale = RCC_D1CFGR_D1HPRE(RCC_D1CFGR_D1HPRE_DIV2),
    // Constrain APB3 below 120MHz by dividing hclk3 by 2.
    .d1cfg_pclk3_prescale = RCC_D1CFGR_D1PPRE(RCC_D1CFGR_D1PPRE_DIV2),
    // Constrain APB1 below 120MHz by dividing HCLK3 by 2.
    .d2cfg_pclk1_prescale = RCC_D2CFGR_D2PPRE1(RCC_D2CFGR_D2PPRE_DIV2),
    // Constrain APB2 below 120MHz by dividing HCLK3 by 2.
    .d2cfg_pclk2_prescale = RCC_D2CFGR_D2PPRE2(RCC_D2CFGR_D2PPRE_DIV2),
    // Constrain APB4 below 120MHz by dividing HCLK3 by 2.
    .d3cfg_pclk4_prescale = RCC_D3CFGR_D3PPRE(RCC_D3CFGR_D3PPRE_DIV2),
  };
  // clang-format on
  rcc_clock_setup_pll(&pll_config);

  // Setup SPI buses to use the HSI clock @ 64MHz for SPI log2 based dividers.
  rcc_set_spi123_clksel(RCC_D2CCIP1R_SPI123SEL_PERCK);  // PERCLK is defaulted to HSI.
  rcc_set_spi45_clksel(RCC_D2CCIP1R_SPI45SEL_HSI);

  // Set CANFD to use PLL1Q set at 80MHz. 
  rcc_set_fdcan_clksel(RCC_D2CCIP1R_FDCANSEL_PLL1Q);
}
```

```c++

The usart code update is one example, but here is a snippet from another one of our drivers that uses the get_clock_freq() interface:

```c++
ErrorReturn Stm32FdCan::SetBitrate(uint32_t bitrate, uint32_t fdrate) noexcept {
  std::lock_guard<std::mutex> lock(mutex_);
  struct can_bittiming nbt {};
  struct can_bittiming fdbt {};
  uint32_t base_freq = rcc_get_clock_freq(RCC_FDCAN1CLK);

  struct can_bittiming_const nbtc FDCAN_NOMINAL_BITTIMING_CONST;
  RET_USR_IF(can_get_bittiming(bitrate, 0, 1, base_freq, &nbtc, &nbt) == false,
             ErrorReturn::Invalid);
  if (fdrate > 0) {
    struct can_bittiming_const fdbtc FDCAN_FD_BITTIMING_CONST;
    RET_USR_IF(can_get_bittiming(fdrate, 0, 1, base_freq, &fdbtc, &fdbt) == false,
               ErrorReturn::Invalid);
  }
```

Reply:
None of these use the enum above?  I'm still concerned about how dense RCC_<something clockish> namespace is, when we used the short form of them for all the clock enabling.  I'm going to try this out, but I'm not excited :|